flake8-comprehensions (C4) - Remove unnecessary list comprehensions in additional contexts that can take arbitrary iterables (C419)

The following code all creates unnecessary list comprehensions

```python
    a = all([i is not None for i in range(10)])
    b = tuple([i * 2 for i in range(10)])
    c = sum([i * 2 for i in range(10)])
    d = ",".join([str(i * 2) for i in range(10)])
```

ruff will only suggest changes on the first line:

```
stuff.py:3:13: C419 Unnecessary list comprehension
  |
2 | def main() -> None:
3 |     a = all([i is not None for i in range(10)])
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ C419
4 |     b = tuple([i * 2 for i in range(10)])
5 |     c = sum([i * 2 for i in range(10)])
  |
  = help: Remove unnecessary list comprehension
```

However all of these are unnecessary list comprehensions. It ought to be able to identify them all. For one, the [docs suggest](https://docs.astral.sh/ruff/rules/unnecessary-comprehension-in-call/) that `sum` is covered, but from the above, it appears to not fully be.

--- a/crates/ruff_linter/resources/test/fixtures/flake8_comprehensions/C409.py
+++ b/crates/ruff_linter/resources/test/fixtures/flake8_comprehensions/C409.py
@@ -24,3 +24,21 @@ tuple((
 t6 = tuple([1])
 t7 = tuple((1,))
 t8 = tuple([1,])
+
+tuple([x for x in range(5)])
+tuple({x for x in range(10)})
+tuple(x for x in range(5))
+tuple([
+    x for x in [1,2,3]
+])
+tuple( # comment
+    [x for x in [1,2,3]]
+)
+tuple([ # comment
+    x for x in range(10)
+])
+tuple(
+    {
+        x for x in [1,2,3]
+    }
+)

--- a/crates/ruff_linter/src/checkers/ast/analyze/expression.rs
+++ b/crates/ruff_linter/src/checkers/ast/analyze/expression.rs
@@ -704,7 +704,9 @@ pub(crate) fn expression(expr: &Expr, checker: &mut Checker) {
                 );
             }
             if checker.enabled(Rule::UnnecessaryLiteralWithinTupleCall) {
-                flake8_comprehensions::rules::unnecessary_literal_within_tuple_call(checker, call);
+                flake8_comprehensions::rules::unnecessary_literal_within_tuple_call(
+                    checker, expr, call,
+                );
             }
             if checker.enabled(Rule::UnnecessaryLiteralWithinListCall) {
                 flake8_comprehensions::rules::unnecessary_literal_within_list_call(checker, call);

--- a/crates/ruff_linter/src/rules/flake8_comprehensions/mod.rs
+++ b/crates/ruff_linter/src/rules/flake8_comprehensions/mod.rs
@@ -45,6 +45,7 @@ mod tests {
         Ok(())
     }
 
+    #[test_case(Rule::UnnecessaryLiteralWithinTupleCall, Path::new("C409.py"))]
     #[test_case(Rule::UnnecessaryComprehensionInCall, Path::new("C419_1.py"))]
     fn preview_rules(rule_code: Rule, path: &Path) -> Result<()> {
         let snapshot = format!(

--- a/crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_literal_within_tuple_call.rs
+++ b/crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_literal_within_tuple_call.rs
@@ -10,13 +10,6 @@ use crate::rules::flake8_comprehensions::fixes;
 
 use super::helpers;
 
-#[derive(Debug, PartialEq, Eq)]
-enum TupleLiteralKind {
-    List,
-    Tuple,
-    ListComp,
-}
-
 /// ## What it does
 /// Checks for `tuple` calls that take unnecessary list or tuple literals as
 /// arguments. In [preview], this also includes unnecessary list comprehensions
@@ -112,7 +105,7 @@ pub(crate) fn unnecessary_literal_within_tuple_call(
     let argument_kind = match argument {
         Expr::Tuple(_) => TupleLiteralKind::Tuple,
         Expr::List(_) => TupleLiteralKind::List,
-        Expr::ListComp(_) => TupleLiteralKind::ListComp,
+        Expr::ListComp(_) if checker.settings.preview.is_enabled() => TupleLiteralKind::ListComp,
         _ => return,
     };
 
@@ -123,57 +116,63 @@ pub(crate) fn unnecessary_literal_within_tuple_call(
         call.range(),
     );
 
-    if argument.is_tuple_expr() | argument.is_list_expr() {
-        // Convert `tuple([1, 2])` to `(1, 2)`
-        diagnostic.set_fix({
-            let elts = match argument {
-                Expr::List(ast::ExprList { elts, .. }) => elts.as_slice(),
-                Expr::Tuple(ast::ExprTuple { elts, .. }) => elts.as_slice(),
-                _ => return,
-            };
-
-            let needs_trailing_comma = if let [item] = elts {
-                SimpleTokenizer::new(
-                    checker.locator().contents(),
-                    TextRange::new(item.end(), call.end()),
-                )
-                .all(|token| token.kind != SimpleTokenKind::Comma)
-            } else {
-                false
-            };
-
-            // Replace `[` with `(`.
-            let elt_start = Edit::replacement(
-                "(".into(),
-                call.start(),
-                argument.start() + TextSize::from(1),
-            );
-            // Replace `]` with `)` or `,)`.
-            let elt_end = Edit::replacement(
-                if needs_trailing_comma {
-                    ",)".into()
+    match argument {
+        Expr::List(ast::ExprList { elts, .. }) | Expr::Tuple(ast::ExprTuple { elts, .. }) => {
+            // Convert `tuple([1, 2])` to `(1, 2)`
+            diagnostic.set_fix({
+                let needs_trailing_comma = if let [item] = elts.as_slice() {
+                    SimpleTokenizer::new(
+                        checker.locator().contents(),
+                        TextRange::new(item.end(), call.end()),
+                    )
+                    .all(|token| token.kind != SimpleTokenKind::Comma)
                 } else {
-                    ")".into()
-                },
-                argument.end() - TextSize::from(1),
-                call.end(),
-            );
-            Fix::unsafe_edits(elt_start, [elt_end])
-        });
-    } else if checker.settings.preview.is_enabled() && argument.is_list_comp_expr() {
-        // Convert `tuple([x for x in range(10)])` to `tuple(x for x in range(10))`
-        let Expr::ListComp(ast::ExprListComp { elt, .. }) = argument else {
-            return;
-        };
-        if any_over_expr(elt, &Expr::is_await_expr) {
-            return;
+                    false
+                };
+
+                // Replace `[` with `(`.
+                let elt_start = Edit::replacement(
+                    "(".into(),
+                    call.start(),
+                    argument.start() + TextSize::from(1),
+                );
+                // Replace `]` with `)` or `,)`.
+                let elt_end = Edit::replacement(
+                    if needs_trailing_comma {
+                        ",)".into()
+                    } else {
+                        ")".into()
+                    },
+                    argument.end() - TextSize::from(1),
+                    call.end(),
+                );
+                Fix::unsafe_edits(elt_start, [elt_end])
+            });
         }
-        diagnostic.try_set_fix(|| {
-            fixes::fix_unnecessary_comprehension_in_call(expr, checker.locator(), checker.stylist())
-        });
-    } else {
-        return;
+
+        Expr::ListComp(ast::ExprListComp { elt, .. }) => {
+            if any_over_expr(elt, &Expr::is_await_expr) {
+                return;
+            }
+            // Convert `tuple([x for x in range(10)])` to `tuple(x for x in range(10))`
+            diagnostic.try_set_fix(|| {
+                fixes::fix_unnecessary_comprehension_in_call(
+                    expr,
+                    checker.locator(),
+                    checker.stylist(),
+                )
+            });
+        }
+
+        _ => return,
     }
 
     checker.diagnostics.push(diagnostic);
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum TupleLiteralKind {
+    List,
+    Tuple,
+    ListComp,
 }

--- a/crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_literal_within_tuple_call.rs
+++ b/crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_literal_within_tuple_call.rs
@@ -10,9 +10,16 @@ use crate::rules::flake8_comprehensions::fixes;
 
 use super::helpers;
 
+#[derive(Debug, PartialEq, Eq)]
+enum TupleLiteralKind {
+    List,
+    Tuple,
+    ListComp,
+}
+
 /// ## What it does
 /// Checks for `tuple` calls that take unnecessary list or tuple literals as
-/// arguments. This includes literals in the form of list or set comprehensions.
+/// arguments. This includes literals in the form of list comprehensions.
 ///
 /// ## Why is this bad?
 /// It's unnecessary to use a list or tuple literal within a `tuple()` call,
@@ -44,32 +51,36 @@ use super::helpers;
 ///
 #[violation]
 pub struct UnnecessaryLiteralWithinTupleCall {
-    literal: String,
+    literal_kind: TupleLiteralKind,
 }
 
 impl AlwaysFixableViolation for UnnecessaryLiteralWithinTupleCall {
     #[derive_message_formats]
     fn message(&self) -> String {
-        let UnnecessaryLiteralWithinTupleCall { literal } = self;
-        if literal == "list" {
-            format!(
-                "Unnecessary `{literal}` literal passed to `tuple()` (rewrite as a `tuple` literal)"
-            )
-        } else {
-            format!(
-                "Unnecessary `{literal}` literal passed to `tuple()` (remove the outer call to `tuple()`)"
-            )
+        let UnnecessaryLiteralWithinTupleCall { literal_kind } = self;
+        match literal_kind {
+            TupleLiteralKind::List => {
+                format!(
+                    "Unnecessary `list` literal passed to `tuple()` (rewrite as a `tuple` literal)"
+                )
+            }
+            TupleLiteralKind::Tuple => {
+                format!("Unnecessary `tuple` literal passed to `tuple()` (remove the outer call to `tuple()`)")
+            }
+            TupleLiteralKind::ListComp => {
+                format!(
+                    "Unnecessary list comprehension passed to `tuple()` (rewrite with generator)"
+                )
+            }
         }
     }
 
     fn fix_title(&self) -> String {
-        let UnnecessaryLiteralWithinTupleCall { literal } = self;
-        {
-            if literal == "list" {
-                "Rewrite as a `tuple` literal".to_string()
-            } else {
-                "Remove outer `tuple` call".to_string()
-            }
+        let UnnecessaryLiteralWithinTupleCall { literal_kind } = self;
+        match literal_kind {
+            TupleLiteralKind::List => "Rewrite as a `tuple` literal.".to_string(),
+            TupleLiteralKind::Tuple => "Remove the outer call to `tuple()`.".to_string(),
+            TupleLiteralKind::ListComp => "Rewrite with generator".to_string(),
         }
     }
 }
@@ -95,16 +106,15 @@ pub(crate) fn unnecessary_literal_within_tuple_call(
         return;
     }
     let argument_kind = match argument {
-        Expr::Tuple(_) => "tuple",
-        Expr::List(_) => "list",
-        Expr::ListComp(_) => "list comprehension",
-        Expr::SetComp(_) => "set comprehension",
+        Expr::Tuple(_) => TupleLiteralKind::Tuple,
+        Expr::List(_) => TupleLiteralKind::List,
+        Expr::ListComp(_) => TupleLiteralKind::ListComp,
         _ => return,
     };
 
     let mut diagnostic = Diagnostic::new(
         UnnecessaryLiteralWithinTupleCall {
-            literal: argument_kind.to_string(),
+            literal_kind: argument_kind,
         },
         call.range(),
     );
@@ -146,10 +156,8 @@ pub(crate) fn unnecessary_literal_within_tuple_call(
             );
             Fix::unsafe_edits(elt_start, [elt_end])
         });
-    } else if argument.is_list_comp_expr() | argument.is_set_comp_expr() {
-        let (Expr::ListComp(ast::ExprListComp { elt, .. })
-        | Expr::SetComp(ast::ExprSetComp { elt, .. })) = argument
-        else {
+    } else if argument.is_list_comp_expr() {
+        let Expr::ListComp(ast::ExprListComp { elt, .. }) = argument else {
             return;
         };
         if any_over_expr(elt, &Expr::is_await_expr) {

--- a/crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_literal_within_tuple_call.rs
+++ b/crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_literal_within_tuple_call.rs
@@ -12,7 +12,7 @@ use super::helpers;
 
 /// ## What it does
 /// Checks for `tuple` calls that take unnecessary list or tuple literals as
-/// arguments.
+/// arguments. This includes literals in the form of list or set comprehensions.
 ///
 /// ## Why is this bad?
 /// It's unnecessary to use a list or tuple literal within a `tuple()` call,
@@ -26,17 +26,22 @@ use super::helpers;
 /// ```python
 /// tuple([1, 2])
 /// tuple((1, 2))
+/// tuple([x for x in range(10)])
 /// ```
 ///
 /// Use instead:
 /// ```python
 /// (1, 2)
 /// (1, 2)
+/// tuple(x for x in range(10))
 /// ```
 ///
 /// ## Fix safety
 /// This rule's fix is marked as unsafe, as it may occasionally drop comments
 /// when rewriting the call. In most cases, though, comments will be preserved.
+/// Moreover, in the case where a comprehension is replaced by a generator,
+/// code behavior may be changed if the iteration has side-effects.
+///
 #[violation]
 pub struct UnnecessaryLiteralWithinTupleCall {
     literal: String,

--- a/crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_literal_within_tuple_call.rs
+++ b/crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_literal_within_tuple_call.rs
@@ -19,7 +19,8 @@ enum TupleLiteralKind {
 
 /// ## What it does
 /// Checks for `tuple` calls that take unnecessary list or tuple literals as
-/// arguments. This includes literals in the form of list comprehensions.
+/// arguments. In [preview], this also includes unnecessary list comprehensions
+/// within tuple calls.
 ///
 /// ## Why is this bad?
 /// It's unnecessary to use a list or tuple literal within a `tuple()` call,
@@ -28,6 +29,10 @@ enum TupleLiteralKind {
 /// If a list literal was passed, then it should be rewritten as a `tuple`
 /// literal. Otherwise, if a tuple literal was passed, then the outer call
 /// to `tuple()` should be removed.
+///
+/// In [preview], this rule also checks for list comprehensions within `tuple()`
+/// calls. If a list comprehension is found, it should be rewritten as a
+/// generator expression.
 ///
 /// ## Examples
 /// ```python
@@ -46,9 +51,8 @@ enum TupleLiteralKind {
 /// ## Fix safety
 /// This rule's fix is marked as unsafe, as it may occasionally drop comments
 /// when rewriting the call. In most cases, though, comments will be preserved.
-/// Moreover, in the case where a comprehension is replaced by a generator,
-/// code behavior may be changed if the iteration has side-effects.
 ///
+/// [preview]: https://docs.astral.sh/ruff/preview/
 #[violation]
 pub struct UnnecessaryLiteralWithinTupleCall {
     literal_kind: TupleLiteralKind,
@@ -69,7 +73,7 @@ impl AlwaysFixableViolation for UnnecessaryLiteralWithinTupleCall {
             }
             TupleLiteralKind::ListComp => {
                 format!(
-                    "Unnecessary list comprehension passed to `tuple()` (rewrite with generator)"
+                    "Unnecessary list comprehension passed to `tuple()` (rewrite as a generator)"
                 )
             }
         }
@@ -78,9 +82,9 @@ impl AlwaysFixableViolation for UnnecessaryLiteralWithinTupleCall {
     fn fix_title(&self) -> String {
         let UnnecessaryLiteralWithinTupleCall { literal_kind } = self;
         match literal_kind {
-            TupleLiteralKind::List => "Rewrite as a `tuple` literal.".to_string(),
-            TupleLiteralKind::Tuple => "Remove the outer call to `tuple()`.".to_string(),
-            TupleLiteralKind::ListComp => "Rewrite with generator".to_string(),
+            TupleLiteralKind::List => "Rewrite as a `tuple` literal".to_string(),
+            TupleLiteralKind::Tuple => "Remove the outer call to `tuple()`".to_string(),
+            TupleLiteralKind::ListComp => "Rewrite as a generator".to_string(),
         }
     }
 }
@@ -119,8 +123,8 @@ pub(crate) fn unnecessary_literal_within_tuple_call(
         call.range(),
     );
 
-    // Convert `tuple([1, 2])` to `(1, 2)`
     if argument.is_tuple_expr() | argument.is_list_expr() {
+        // Convert `tuple([1, 2])` to `(1, 2)`
         diagnostic.set_fix({
             let elts = match argument {
                 Expr::List(ast::ExprList { elts, .. }) => elts.as_slice(),
@@ -156,7 +160,8 @@ pub(crate) fn unnecessary_literal_within_tuple_call(
             );
             Fix::unsafe_edits(elt_start, [elt_end])
         });
-    } else if argument.is_list_comp_expr() {
+    } else if checker.settings.preview.is_enabled() && argument.is_list_comp_expr() {
+        // Convert `tuple([x for x in range(10)])` to `tuple(x for x in range(10))`
         let Expr::ListComp(ast::ExprListComp { elt, .. }) = argument else {
             return;
         };

--- a/crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_literal_within_tuple_call.rs
+++ b/crates/ruff_linter/src/rules/flake8_comprehensions/rules/unnecessary_literal_within_tuple_call.rs
@@ -1,10 +1,12 @@
 use ruff_diagnostics::{AlwaysFixableViolation, Diagnostic, Edit, Fix};
 use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::helpers::any_over_expr;
 use ruff_python_ast::{self as ast, Expr};
 use ruff_python_trivia::{SimpleTokenKind, SimpleTokenizer};
 use ruff_text_size::{Ranged, TextRange, TextSize};
 
 use crate::checkers::ast::Checker;
+use crate::rules::flake8_comprehensions::fixes;
 
 use super::helpers;
 
@@ -68,7 +70,11 @@ impl AlwaysFixableViolation for UnnecessaryLiteralWithinTupleCall {
 }
 
 /// C409
-pub(crate) fn unnecessary_literal_within_tuple_call(checker: &mut Checker, call: &ast::ExprCall) {
+pub(crate) fn unnecessary_literal_within_tuple_call(
+    checker: &mut Checker,
+    expr: &Expr,
+    call: &ast::ExprCall,
+) {
     if !call.arguments.keywords.is_empty() {
         return;
     }
@@ -86,6 +92,8 @@ pub(crate) fn unnecessary_literal_within_tuple_call(checker: &mut Checker, call:
     let argument_kind = match argument {
         Expr::Tuple(_) => "tuple",
         Expr::List(_) => "list",
+        Expr::ListComp(_) => "list comprehension",
+        Expr::SetComp(_) => "set comprehension",
         _ => return,
     };
 
@@ -97,41 +105,57 @@ pub(crate) fn unnecessary_literal_within_tuple_call(checker: &mut Checker, call:
     );
 
     // Convert `tuple([1, 2])` to `(1, 2)`
-    diagnostic.set_fix({
-        let elts = match argument {
-            Expr::List(ast::ExprList { elts, .. }) => elts.as_slice(),
-            Expr::Tuple(ast::ExprTuple { elts, .. }) => elts.as_slice(),
-            _ => return,
-        };
+    if argument.is_tuple_expr() | argument.is_list_expr() {
+        diagnostic.set_fix({
+            let elts = match argument {
+                Expr::List(ast::ExprList { elts, .. }) => elts.as_slice(),
+                Expr::Tuple(ast::ExprTuple { elts, .. }) => elts.as_slice(),
+                _ => return,
+            };
 
-        let needs_trailing_comma = if let [item] = elts {
-            SimpleTokenizer::new(
-                checker.locator().contents(),
-                TextRange::new(item.end(), call.end()),
-            )
-            .all(|token| token.kind != SimpleTokenKind::Comma)
-        } else {
-            false
-        };
-
-        // Replace `[` with `(`.
-        let elt_start = Edit::replacement(
-            "(".into(),
-            call.start(),
-            argument.start() + TextSize::from(1),
-        );
-        // Replace `]` with `)` or `,)`.
-        let elt_end = Edit::replacement(
-            if needs_trailing_comma {
-                ",)".into()
+            let needs_trailing_comma = if let [item] = elts {
+                SimpleTokenizer::new(
+                    checker.locator().contents(),
+                    TextRange::new(item.end(), call.end()),
+                )
+                .all(|token| token.kind != SimpleTokenKind::Comma)
             } else {
-                ")".into()
-            },
-            argument.end() - TextSize::from(1),
-            call.end(),
-        );
-        Fix::unsafe_edits(elt_start, [elt_end])
-    });
+                false
+            };
+
+            // Replace `[` with `(`.
+            let elt_start = Edit::replacement(
+                "(".into(),
+                call.start(),
+                argument.start() + TextSize::from(1),
+            );
+            // Replace `]` with `)` or `,)`.
+            let elt_end = Edit::replacement(
+                if needs_trailing_comma {
+                    ",)".into()
+                } else {
+                    ")".into()
+                },
+                argument.end() - TextSize::from(1),
+                call.end(),
+            );
+            Fix::unsafe_edits(elt_start, [elt_end])
+        });
+    } else if argument.is_list_comp_expr() | argument.is_set_comp_expr() {
+        let (Expr::ListComp(ast::ExprListComp { elt, .. })
+        | Expr::SetComp(ast::ExprSetComp { elt, .. })) = argument
+        else {
+            return;
+        };
+        if any_over_expr(elt, &Expr::is_await_expr) {
+            return;
+        }
+        diagnostic.try_set_fix(|| {
+            fixes::fix_unnecessary_comprehension_in_call(expr, checker.locator(), checker.stylist())
+        });
+    } else {
+        return;
+    }
 
     checker.diagnostics.push(diagnostic);
 }

